precompiles/assets/types.go

```diff
@@ -32,27 +35,132 @@ func (p Precompile) GetDepositWithdrawParamsFromInputs(ctx sdk.Context, args []i
+	}
+	clientChain.AddressLength = uint32(addressLength)
+
+	name, ok := args[2].(string)
+	if !ok {
+		return nil, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 2, "string", args[2])
+	}
+	if name == "" || len(name) > types.MaxChainTokenNameLength {
+		return nil, xerrors.Errorf(exocmn.ErrInvalidNameLength, name, len(name), types.MaxChainTokenNameLength)
+	}
+	clientChain.Name = name
+
+	metaInfo, ok := args[3].(string)
+	if !ok {
+		return nil, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 3, "string", args[2])
+	}
+	if metaInfo == "" || len(metaInfo) > types.MaxChainTokenMetaInfoLength {
+		return nil, xerrors.Errorf(exocmn.ErrInvalidMetaInfoLength, metaInfo, len(metaInfo), types.MaxChainTokenMetaInfoLength)
+	}
+	clientChain.MetaInfo = metaInfo
+
+	signatureType, ok := args[4].(string)
+	if !ok {
+		return nil, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 4, "string", args[4])
+	}
+	clientChain.SignatureType = signatureType
+
+	return &clientChain, nil
+}
+
+func (p Precompile) TokenFromInputs(ctx sdk.Context, args []interface{}) (types.AssetInfo, error) {
+	if len(args) != 6 {
+		return types.AssetInfo{}, xerrors.Errorf(cmn.ErrInvalidNumberOfArgs, 6, len(args))
+	}
+	asset := types.AssetInfo{}
+	clientChainID, ok := args[0].(uint32)
+	if !ok {
+		return types.AssetInfo{}, fmt.Errorf(exocmn.ErrContractInputParaOrType, 0, "uint32", args[0])
+	}
+	asset.LayerZeroChainID = uint64(clientChainID)
+	info, err := p.assetsKeeper.GetClientChainInfoByIndex(ctx, asset.LayerZeroChainID)
+	if err != nil {
+		return types.AssetInfo{}, err
+	}
+	clientChainAddrLength := info.AddressLength
+
+	assetAddr, ok := args[1].([]byte)
+	if !ok || assetAddr == nil {
+		return types.AssetInfo{}, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 1, "[]byte", args[1])
+	}
+	if uint32(len(assetAddr)) < clientChainAddrLength {
+		return types.AssetInfo{}, xerrors.Errorf(exocmn.ErrInvalidAddrLength, len(assetAddr), clientChainAddrLength)
+	}
+	asset.Address = hexutil.Encode(assetAddr[:clientChainAddrLength])
+
+	decimal, ok := args[2].(uint8)
+	if !ok {
+		return types.AssetInfo{}, fmt.Errorf(exocmn.ErrContractInputParaOrType, 2, "uint8", args[2])
+	}
+	asset.Decimals = uint32(decimal)
+
+	tvlLimit, ok := args[3].(*big.Int)
+	if !ok {
+		return types.AssetInfo{}, fmt.Errorf(exocmn.ErrContractInputParaOrType, 3, "*big.Int", args[3])
+	}
+	asset.TotalSupply = sdkmath.NewIntFromBigInt(tvlLimit)
+
+	name, ok := args[4].(string)
+	if !ok {
+		return types.AssetInfo{}, fmt.Errorf(exocmn.ErrContractInputParaOrType, 4, "string", args[4])
+	}
+	if name == "" || len(name) > types.MaxChainTokenNameLength {
+		return types.AssetInfo{}, xerrors.Errorf(exocmn.ErrInvalidNameLength, name, len(name), types.MaxChainTokenNameLength)
+	}
+	asset.Name = name
+
+	metaInfo, ok := args[5].(string)
+	if !ok {
+		return types.AssetInfo{}, fmt.Errorf(exocmn.ErrContractInputParaOrType, 5, "string", args[5])
+	}
+	if metaInfo == "" || len(metaInfo) > types.MaxChainTokenMetaInfoLength {
```

Comment:
same as above comment it seems reasonable to allow metaInfo being empty

Reply:
It seems better not to allow the metaInfo to be empty if we provide this field to describe the client chain. This can ensure detailed and clearer information about the client chain for all users. Of course, this is a minor issue that won't have a significant impact no matter how it's handled.

Reply:
ok, I'd better add this check in contract as well